Check if media commands are actually applicable

bjw-s opened the pull request:
## Description:
- Explicitly allow `stop` and `play` on radio streams
- Disallow media commands when the playlist is empty (doesn't make any sense to `play`, `stop` or `pause` an empty playlist)
- Check if command is supported when calling `turn_on` and `turn_off`

Implementing these checks prevents the UPnP Error 701 (Transition not available)

**Related issue (if applicable):** fixes #4816

## Checklist:

If the code communicates with devices, web services, or third-party tools:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

bjw-s committed:
Check if media commands are actually applicable

- Explicitly allow ‘stop’ and ‘play’ on radio streams
- Disallow media commands when the playlist is empty
- Check if command is supported when calling `turn_on` and `turn_off`

balloob commented on homeassistant/components/media_player/sonos.py:
```diff
@@ -608,6 +613,13 @@ def update(self):
             else:
                 playlist_size = int(playlist_size)
 
+            if playlist_size is None or playlist_size == 0:
```
I am kind of reluctant to include business logic. What does Sonos do when you call play and there is no playlist size?

Juggels replied:
Understandable, as I felt the same way at first.

The Sonos controller just says "No music", and the play/pause button does nothing on the surface. From a UPnP standpoint, it's valid for the SoCo package to throw the 701 error as playing or pausing, etc. on an empty playlist is an invalid state transition (because there is nothing to play or pause). So the options were to try/catch the 701 error everywhere, or control the play/pause support based on playlist size.

balloob replied:
If we start catching 701 errors we wouldn't have to implement any business logic and just make sure we don't log them. That sounds more maintainable. Is there any added value from going down the business logic route?

Juggels replied:
There is not necessarily any added value I think. I figured this would be a bit more clean, as this way the logic for when play/pause/etc is supported is explicit. Also I wasn't not sure where I would best implement the catch statements. I guess that would be in the `self.media_pause()` (etc) functions?

balloob replied:
Can you do something similar to the `soco_error` decorator? https://github.com/Juggels/home-assistant/blob/db8414048ee85cb63ac388b3f8a4b571a2f9074f/homeassistant/components/media_player/sonos.py#L248-L259

Juggels replied:
I'll give it a go :)

btw, just noticed that the macOS Sonos Controller app also disables the play/pause button upon an empty playlist (apparently the iOS app doesn't, that's where I checked earlier)

bjw-s committed:
Suppress UPnP error 701 on media commands

Juggels commented:
Processed requested changes in latest commit

balloob commented on homeassistant/components/media_player/sonos.py:
```diff
@@ -255,10 +258,37 @@ def wrapper(*args, **kwargs):
+                else:
+                    raise
+
+            _SOCO_SERVICES_LOGGER.disabled = False
```
Add this to the `finally` block from try…catch so that it's guaranteed to enable logging after it runs.

balloob commented on homeassistant/components/media_player/sonos.py:
```diff
@@ -255,10 +258,37 @@ def wrapper(*args, **kwargs):
+
+            returnval = None
+            try:
+                returnval = funct(*args, **kwargs)
```
you can just return the value once you have added the finally block 👍

balloob commented on homeassistant/components/media_player/sonos.py:
```diff
@@ -255,10 +258,37 @@ def wrapper(*args, **kwargs):
+            try:
+                returnval = funct(*args, **kwargs)
+            except SoCoUPnPException as err:
+                if errorcodes is not None and err.error_code in errorcodes:
```
There is no reason to support an errorcodes being `None`

bjw-s committed:
Clean up soco_filter_upnperror

Clean up soco_filter_upnperror and fix small bug in support_previous_track determination

Juggels commented:
Changes processed 😄
